File: lc/Tree/古城-Lower Common Ancestor(公共祖先)/236. Lowest Common Ancestor of a Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    if root == nil {
        return nil
    }

    if root.Val == p.Val || root.Val == q.Val {
        return root
    }

    l := lowestCommonAncestor(root.Left, p, q)
    r := lowestCommonAncestor(root.Right, p, q)

    if l != nil && r != nil {
        return root
    }
    if l != nil {
        return l
    }
    if r != nil {
        return r
    }
    return nil
}

File: lc/Tree/古城-Lower Common Ancestor(公共祖先)/236. Lowest Common Ancestor of a Binary Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function lowestCommonAncestor(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {
	if (root === null) {
        return null;
    }

    if (root.val === p.val || root.val === q.val) {
        return root;
    }

    let l = lowestCommonAncestor(root.left, p, q)
    let r = lowestCommonAncestor(root.right, p, q)

    if (l !== null && r !== null) {
        return root;
    }
    if (l !== null) {
        return l
    }
    if (r !== null) {
        return r
    }
    return null
};